Add support for FreeBSD

Kintaro opened the pull request:
### Description

This change adds support for FreeBSD. The change is rather straightforward as FreeBSD also supports uinput and evdev, hence it mostly just requires adding the appropriate `ifdef`s and a slight change in include paths.

Currently daily driving it on my laptops without any issues.

### Checklist

- [X] I've read [CONTRIBUTING.md](https://github.com/kmonad/kmonad/blob/master/CONTRIBUTING.md)
- [X] I've also appended my changes to the `CHANGELOG.md` if applicable

Commit:
Add support for FreeBSD

jokesper commented on c_src/keyio.c:
```diff
@@ -3,8 +3,13 @@
+#if __linux__
 #include <linux/input.h>
 #include <linux/uinput.h>
+#elif __FreeBSD__
```
In my attempt to implement this, I needed:
```suggestion
#elif __FreeBSD__
#include <sys/ioctl.h>
```
I assume I didn't put that there without purpose though I'm not sure.

jokesper reviewed (commented):
I found my own try at this from some time back: https://github.com/jokesper/kmonad/commit/d3e9156f30fdd319c5a7a78c2052902bfec26fca

Now there are just the remaining problems with TTY and running in a CI to make sure it doesn't break in the future (See [GitHub CI issue](https://github.com/actions/runner/issues/385)).

I think we can do without an CI in the mean time, tough depending on how KMonad on the tty is broken we may need anything from an faq entry to a note in installation instructions

Commit:
Close device whhen switching to TTY

While on TTY, the device is closed to allow the TTY to read from the keyboard.
A workaround to get kmonad working on FreeBSD.

jokesper commented:
Ok running with `kmonad -ldebug` it seems kmonad gets the evdev events and tries to output uinput events.
I couldn't find the proper way to generate key events on freebsd. Maybe `uhid` or `wskbd` could work. X11 also supports this though then we run into the same problems we have on windows

Furthermore the dependency `hinotify` does not compile for me (forgot the `libinotify` dependency) This should be documented if we document freebsd install instructions.

Idealy we would put the tty logic in a custom freebsd sink but due to the code structure this seems like a lot of work.

For code structure we may want to add more customizability to the linux sources, so we can have freebsd sources, which modify the lsRead config. Furthermore the checking delay should probably be customizable

jokesper commented:
I think I found the more low level driver. We may probably want to use vkbd, uhid or cuse4bsd-kmod.

Based on the [wiki entry on uhidd](https://wiki.freebsd.org/uhidd) if we want to use vkbd we probably need all pressable buttons and generate a fitting scan code translation table from a config or generate one on the fly.

That being said I think the workaround is fine for now. If I have a bit more time I will write some documentation and structure change (splitting the workaround into seperate files. c_src/keyio_freebsd and the devicesource.hs) and send them to you to review (PR on your fork. Is this how it works on GitHub?) since I don't want to push some bigger change into the PR without your approval. Of course if you want to finish those details yourself you are also welcome.

jokesper commented on changelog.md:
```diff
@@ -20,6 +20,7 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0)
   Buttons binded to it will never trigger. (#992)
 - Using `:ignore-missing` in `device-file` you can wait for the device to be
   connected before starting and allow reconnecting in case of disconnect. (#1017)
+- Added support for FreeBSD
```
Hmm, you are right. It tested it a while back and it seemed broken. interesting.

jokesper commented:
whoops kind of forgot about this. sorry. Sadly I don't think I have the time to fix it soon. I think having a proper implementation would be better. I don't like how the workaround touches so much of the linux part (i.e. extract out). Also the timeout solution for reading keys is slightly weird. this means keys may be pressed indefinitely and be captured. maybe a check for each keystroke would be better. I don't know how expensive the is_vt call is.
